pkg/stream/connpool/msgconnpool/connpool.go

```diff
@@ -0,0 +1,164 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package msgconnpool
+
+import (
+	"errors"
+	"math"
+	"mosn.io/mosn/pkg/log"
+	"sync"
+	"sync/atomic"
+
+	v2 "mosn.io/mosn/pkg/config/v2"
+	"mosn.io/mosn/pkg/upstream/cluster"
+
+	"mosn.io/api"
+	"mosn.io/mosn/pkg/types"
+	"mosn.io/pkg/buffer"
+)
+
+// the state for message conn
+type State int
+
+const (
+	Available State = iota
+	Connecting
+	Destroyed
+)
+
+// Connection is abstraction for message conn
+// which can be configured to auto reconnect when the under conn is broken
+type Connection interface {
+	Write(buf ...buffer.IoBuffer) error // write to current connection
+	State() State                       // get the conn detailed state
+	Destroy()                           // destroy the current conn
+}
+
+type connpool struct {
+	client        *activeClient
+	host          types.Host
+	keepalive     KeepAlive
+	connectingMux sync.Mutex // all the connect behavior should be serial
+
+	autoReconnectWhenClose bool
+	connTryTimes           int
+	destroyed              uint64
+
+	// when connection change, readFilter need to change, use creator to create new filter chain
+	getReadFilterAndKeepalive func() ([]api.ReadFilter, KeepAlive)
+}
+
+// NewConn returns a simplified connpool
+func NewConn(hostAddr string, connectTryTimes int,
+	getReadFilterAndKeepalive func() ([]api.ReadFilter, KeepAlive), autoReconnectWhenClose bool) Connection {
+	// use connData addr as cluster name, for the count of metrics
+	cl := basicCluster(hostAddr, []string{hostAddr})
+	host := cluster.NewSimpleHost(cl.Hosts[0], cluster.NewCluster(cl).Snapshot().ClusterInfo())
+
+	// if user configure this to -1, then retry is unlimited
+	if connectTryTimes == -1 {
+		connectTryTimes = math.MaxInt32
+	}
+
+	p := &connpool{
+		host:                      host,
+		autoReconnectWhenClose:    autoReconnectWhenClose,
+		connTryTimes:              connectTryTimes,
+		getReadFilterAndKeepalive: getReadFilterAndKeepalive,
+	}
+
+	p.initActiveClient()
+
+	return p
+}
+
+func (p *connpool) Host() types.Host {
+	return p.host
+}
+
+///////////// Connection interface start
+
+// Write to client
+func (p *connpool) Write(buf ...buffer.IoBuffer) error {
+	if h, state := p.getConnAndState(); state == Available {
+		return h.Connection.Write(buf...)
+	}
+	return errors.New("[connpool] connection not ready, host" + p.Host().AddressString())
+}
+
+// State current available to send request
+func (p *connpool) State() State {
+	_, state := p.getConnAndState()
+	return state
+}
+
+// Destroy the pool
+func (p *connpool) Destroy() {
```

Comment:
Everything looks okay,but the func Destroy.I can find two unexpected scenes as followings:
1. a panic in the scene:
Calling Destroy before reaching the `ac.connData.Store(&createConnData)` in func initConnectionLocked,that's to say we would get nil pointer with certain probability because of nil connData in activeClient?

2. Same as item 1, calling func Destroy and func initConnectionLocked concurrently, the scene as closing connection executed in func Destroy before reaching `ac.setHeartBeater(keepalive)` might happen, which would cause heartbeater cant capture the close event? 

And we could avoid if we make sure that:func Destroy only called after func initConnectionLocked called? Using a cas lock with timeout or any other way could promise it?

That's all, check it please.. :) 👍

Reply:
1. after the first connect operation, the connData will never be nil, because the first connect is synchronized process until this field is set, if connect is failed, it will still store an object in this connData.

2. after setHeartBeater, the destroyed flag will be checked again

```go
		ac.setHeartBeater(keepalive)
	}

	// two scenes
	//   1. destroyed == true, there is no problem
	//   2. destroyed == false, concurrent with Destroy
	//      Destroy will Close this connection for us
	if atomic.LoadUint64(&ac.pool.destroyed) == 1 {
		ac.clearHeartBeater()
		ac.getConnData().Connection.Close(api.NoFlush, api.LocalClose)
		return
	}
```

Reply:
> 1. when you call NewConn, the first connect is a synchronized procedure, and whether the connect succeeds or not, we both set this field to a non-nil value,  meaning that once NewConn() returns, connData is ensured not nil.
> 2. after setHeartBeater, the destroyed flag will be checked again
> 
> ```go
> 		ac.setHeartBeater(keepalive)
> 	}
> 
> 	// two scenes
> 	//   1. destroyed == true, there is no problem
> 	//   2. destroyed == false, concurrent with Destroy
> 	//      Destroy will Close this connection for us
> 	if atomic.LoadUint64(&ac.pool.destroyed) == 1 {
> 		ac.clearHeartBeater()
> 		ac.getConnData().Connection.Close(api.NoFlush, api.LocalClose)
> 		return
> 	}
> ```

Thanks for checking..
I see item 1 is okay,but item 2 still exist?
Referring to `connection` :
```
func (c *connection) Close(ccType api.ConnectionCloseType, eventType api.ConnectionEvent) error {
	defer func() {
		if p := recover(); p != nil {
			log.DefaultLogger.Errorf("[network] [close connection] panic %v\n%s", p, string(debug.Stack()))
		}
	}()

	if ccType == api.FlushWrite {
		c.Write(buffer.NewIoBufferEOF())
		return nil
	}

	if !atomic.CompareAndSwapUint32(&c.closed, 0, 1) {
		return nil
	}

	// connection failed in client mode
	if c.rawConnection == nil || reflect.ValueOf(c.rawConnection).IsNil() {
		return nil
	}

	...
}
```
Cant `close` connection twice,and `OnEvent` wouldnt called as callback of `close` ?

Reply:
![image](https://user-images.githubusercontent.com/384546/85205278-ff47bf80-b34c-11ea-8ab9-cbdd6b959b42.png)